fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlanner.java

```diff
@@ -1424,14 +1426,38 @@ private RecordQueryPlan planScan(@Nonnull CandidateScan candidateScan,
             }
             plan = new RecordQueryScanPlan(possibleTypes, new Type.Any(), candidateScan.planContext.commonPrimaryKey, scanComparisons, candidateScan.reverse, strictlySorted);
         } else {
-            plan = new RecordQueryIndexPlan(candidateScan.index.getName(), candidateScan.planContext.commonPrimaryKey, indexScanComparisons, getConfiguration().getIndexFetchMethod(), candidateScan.reverse, strictlySorted);
+            final RecordQueryIndexPlan.FetchIndexRecords fetchIndexRecords = resolveFetchIndexRecords(candidateScan);
+            // If this is a regular fetch using the primary key, we can opt to use the configured fetch method,
+            // if, however, this fetch fetches synthetic constituents, we must do it at the client (at this point).
+            final IndexFetchMethod indexFetchMethod =
+                    fetchIndexRecords == RecordQueryIndexPlan.FetchIndexRecords.PRIMARY_KEY
+                    ? getConfiguration().getIndexFetchMethod()
+                    : IndexFetchMethod.SCAN_AND_FETCH;
+            plan = new RecordQueryIndexPlan(candidateScan.index.getName(), candidateScan.planContext.commonPrimaryKey, indexScanComparisons, indexFetchMethod, fetchIndexRecords, candidateScan.reverse, strictlySorted);
             possibleTypes = getPossibleTypes(candidateScan.index);
         }
         // Add a type filter if the query plan might return records of more types than the query specified
         plan = addTypeFilterIfNeeded(candidateScan, plan, possibleTypes);
         return plan;
     }
 
+    @Nonnull
+    RecordQueryIndexPlan.FetchIndexRecords resolveFetchIndexRecords(@Nonnull CandidateScan candidateScan) {
```

Comment:
I'm not sure I follow the logic in this method. I take it, it's something like:

1. If there are no synthetic record types in the meta-data, then fetch by primary key (okay)
2. If there are synthetic record types and all of the types in the query are synthetic, then fetch by synthetic constituents (okay)
3. Otherwise, if there are no types being queried (I guess in which case the fetch method doesn't matter? I'm not sure I follow) or at least one queried type is not synthetic, fetch by primary key

It's mainly that third condition that I'm not sure I follow. Should this throw an error if there are both synthetic and non-synthetic record types in a single query (or handle it somehow differently)? And what does an empty list of queried record types signify (and thus why is it okay to fetch by primary key here)?

Reply:
No types (empty list) being queried is a special case of query everything. I will change the universal index maintainers to ignore synthetic types, so that in this case no synthetic records will ever be returned. Will add comment.